README.md
# Symmetric File

A gem for reading and editing files encrypted with symmetric encryption.

##  Why?

This gem is intended to be a low-hassle, yet secure way to encrypt sensitive data in a git repo. The
encrypted file can be stored safely and any unauthorised access to the repository won't result in
leaked secrets.

It's primarily designed for encrypting configuration files and SSL certificates, but it's very
capable of encrypting any text file.

The files are enrypted using AES-256 in CBC mode, and use HMAC for message authentication. All encryption
and signing is performed using the ruby openssl bindings available in the standard library.

symmetric\_file is based on many of the ideas in the [dotgpg gem](https://rubygems.org/gems/dotgpg) by
Conrad Irwin. Smaller teams may not need the complexity of gnupg and may be comfortable with the security
provided by symmetric encryption and a shared password.

## Installation

Like all gems:

    gem install symmetric_file

.. or just add it to your project Gemfile.

    gem "symmetric_file"

## Usage

This is a command line application. To create a new encypted file, or edit an existing one:

    bundle exec symmetric-file edit config/secrets.yml.enc

To decrypt a file to stdout:

    bundle exec symmetric-file cat config/secrets.yml.enc

To decrypt a file and redirect the output to an unencrypted file:

    bundle exec symmetric-file cat config/secrets.yml.enc > config/secrets.yml

## Git Integration

By default, files encrypted with symmetric\_file don't play well with Git. They're
binary data, so git will not attempt to merge conflicting changes.

symmetric\_file has some helpers that make it possible to merge changes to an encrypted
file, however it requires some configuration of your git repo.

First, add a .gitattributes file to the repo with the following content. This file
can be tracked in git and shared with collaborators.

```
*.aes diff=aes merge=aes
```

Second, edit .git/config and add the following content. This file is specific to the
local repo and is not shared with collaborators.

```
[diff "aes"]
  textconv = bundle exec symmetric-file cat
[merge "aes"]
  name = symmetric-file merge driver
  driver = "bundle exec symmetric-file merge %O %A %B"
```

This will lead to the following behaviour:

A `git diff` that includes changes to an encrypted file will prompt for the
passphrase and then display a diff of the decrypted content.

A `git merge` that involves changes to an encrypted file prompt for the
passphrase. If an automated merge of the decrypted content is possible, there's
nothing left to do. If an automated merge wasn't possible, then the file will
be re-encrypted with merge conflict markers and they can be resolved manually
before commiting the merge.


CHANGELOG
v0.1.0
* initial release


spec/spec_helper.rb
require 'symmetric_file'

RSpec.configure do |config|
  config.disable_monkey_patching!

  config.order = :random
end


spec/symmetric_file/aes_spec.rb
RSpec.describe SymmetricFile::Aes do
  let(:key) { "1234" }
  let(:aes) { SymmetricFile::Aes.new(key: key) }

  describe "Cleartext Round trip" do
    let(:cleartext) { "Test" }

    context "with a short cleartext" do
      let(:encrypted_msg) { aes.encrypt(cleartext)} 
      let(:decrypted_msg) { aes.decrypt(encrypted_msg) }

      it "survives" do
        expect(decrypted_msg).to eq cleartext
      end
    end
    context "with a long cleartext" do
      let(:cleartext) { SecureRandom.hex(5000) }
      let(:encrypted_msg) { aes.encrypt(cleartext)} 
      let(:decrypted_msg) { aes.decrypt(encrypted_msg) }

      it "survives" do
        expect(decrypted_msg).to eq cleartext
      end
    end
  end

  describe "#encrypt" do
    let(:cleartext) { "Test" }
    let(:encrypted_msg) { aes.encrypt(cleartext)} 

    context "when encrypting 'Test'" do
      it "is marked as version 3 of RNCryptor" do
        expect(encrypted_msg[0,1]).to eq "\x03"
      end
      it "is exactly 82 bytes" do
        expect(encrypted_msg.bytesize).to eq 82
      end
      it "is marked as binary" do
        expect(encrypted_msg.encoding).to eq Encoding.find("binary")
      end
    end
    context "when the same cleartext is encrypted twice" do
      let(:encrypted_msg_one) { aes.encrypt(cleartext)} 
      let(:encrypted_msg_two) { aes.encrypt(cleartext)} 

      it "encrypts differently each time" do
        expect(encrypted_msg_one).to_not eq encrypted_msg_two
      end
    end
  end

  describe "#decrypt" do
    context "when the input data is 66 bytes" do
      let(:encrypted_msg) { "a" * 66 }
      
      it "raises InputError" do
        expect {
          aes.decrypt(encrypted_msg)
        }.to raise_error(SymmetricFile::InputError, "Encrypted data too short")
      end
    end
    context "when the input data doesn't indicate v3 of RNCryptor" do
      let(:encrypted_msg) { "\x02" + ("a" * 66) }
      
      it "raises InputError" do
        expect {
          aes.decrypt(encrypted_msg)
        }.to raise_error(SymmetricFile::InputError, "Only version 3 data can be decrypted")
      end
    end
    context "when the input data doesn't have a valid HMAC" do
      let(:encrypted_msg) { "\x03" + ("a" * 67) }
      
      it "raises InputError" do
        expect {
          aes.decrypt(encrypted_msg)
        }.to raise_error(SymmetricFile::InputError, "HMAC validation failed. The passphrase may be incorrect or the encrypted message may have been tampered with")
      end
    end
    context "when the provided key is wrong" do
      let(:encrypted_msg) { aes.encrypt("Test") }
      let(:aes_with_wrong_key) { SymmetricFile::Aes.new(key: "000") }
      
      it "raises InputError" do
        expect {
          aes_with_wrong_key.decrypt(encrypted_msg)
        }.to raise_error(SymmetricFile::InputError, "HMAC validation failed. The passphrase may be incorrect or the encrypted message may have been tampered with")
      end
    end
    context "with valid v3 message" do
      let(:encrypted_hex) {
        "0301835b93e734143340ca8b55fc77865be906abe119073b77d5bc46" +
        "1fcc8bc8aea42fde3eb01b33bd3b54f2d58aaaef7747d24e1bde83aa" +
        "b5f81d7e68e3e2ba6c4f1420b638faea3d6dec7c801345d5bc059289" +
        "f52b4d030786fc11e22a3939efd7c88a6cad3e23a9fc87e6bbfbc389" +
        "01525b2ef7384045923260b3928a5bedbf7b"
      }
      let(:encrypted_msg) { [encrypted_hex].pack("H*") }
      let(:aes) { SymmetricFile::Aes.new(key: "P@ssw0rd!") }
      
      it "returns a message" do
        expect(aes.decrypt(encrypted_msg)).to eq "Hello, World! Let's use a few blocks with a longer sentence."
      end
    end
    context "with valid v3 message" do
      let(:encrypted_hex) {
        "0301835b93e734143340ca8b55fc77865be906abe119073b77d5bc46" +
        "1fcc8bc8aea42fde3eb01b33bd3b54f2d58aaaef7747d24e1bde83aa" +
        "b5f81d7e68e3e2ba6c4f1420b638faea3d6dec7c801345d5bc059289" +
        "f52b4d030786fc11e22a3939efd7c88a6cad3e23a9fc87e6bbfbc389" +
        "01525b2ef7384045923260b3928a5bedbf7b"
      }
      let(:encrypted_msg) { [encrypted_hex].pack("H*") }
      let(:aes) { SymmetricFile::Aes.new(key: "P@ssw0rd!") }
      
      it "returns a message" do
        expect(aes.decrypt(encrypted_msg)).to eq "Hello, World! Let's use a few blocks with a longer sentence."
      end
    end

  end

end


spec/symmetric_file/cat_command_spec.rb
RSpec.describe SymmetricFile::CatCommand do
  let(:key) { "1234" }
  let(:stdout) { double($stdout, puts: true) }
  let(:command) { SymmetricFile::CatCommand.new(key: key, output: stdout) }

  describe "#run" do
    let(:aes) { instance_double(SymmetricFile::Aes)}
    before do
      allow(SymmetricFile::Aes).to receive(:new) { aes }
    end

    context "when the input data can be decrypted" do
      before do
        allow(aes).to receive(:decrypt) { "1234" } 
      end
      it "prints the cleartext to output" do
        command.run("encrypted")
        expect(stdout).to have_received(:puts).with("1234")
      end
    end
    context "when the input data can't be decrypted" do
      before do
        allow(aes).to receive(:decrypt).and_raise(SymmetricFile::InputError)
      end
      it "raises an exception" do
        expect {
          command.run("encrypted")
        }.to raise_error(SymmetricFile::InputError)
      end
    end
  end

end


lib/symmetric_file/cli.rb
require 'tempfile'

module SymmetricFile
  class Cli

    def cat(encrypted_path)
      SymmetricFile::CatCommand.new(key: key).run(read_file(encrypted_path))
    end

    def edit(encrypted_path)
      SymmetricFile::EditCommand.new(
        key: key,
        editor: detect_editor,
      ).run(encrypted_path)
    end

    def merge(mine_path, old_path, yours_path)
      SymmetricFile::MergeCommand.new(key: key).run(mine_path, old_path, yours_path)
    end

    private

    # This method copied directly from Pry and is
    # Copyright (c) 2013 John Mair (banisterfiend)
    # https://github.com/pry/pry/blob/master/LICENSE
    def detect_editor
      configured = ENV["VISUAL"] || ENV["EDITOR"] || guess_editor
      configured = configured.dup
      case configured
      when /^mate/, /^subl/
        configured << " -w"
      when /^[gm]vim/
        configured << " --nofork"
      when /^jedit/
        configured << " -wait"
      end

      configured
    end

    def guess_editor
      %w(subl sublime-text sensible-editor editor mate nano vim vi open).detect do |editor|
        system("command -v #{editor} > /dev/null 2>&1")
      end
    end

    def read_file(path)
      File.binread(path)
    rescue Errno::ENOENT
      raise SymmetricFile::InputError, "file '#{path}' not found"
    end

    def key
      @key ||= read_key("Enter key:")
    end

    def read_input(prompt)
      $stderr.print prompt
      $stderr.flush
      $stdin.readline.strip
    end

    def read_key(prompt)
      `stty -echo`
      read_input prompt
    ensure
      $stderr.print "\n"
      `stty echo`
    end
  end
end


lib/symmetric_file/edit_command.rb
require 'shellwords'

module SymmetricFile
  class EditCommand
    def initialize(key: nil, editor: nil)
      @editor = editor || "vim"
      @cipher = SymmetricFile::Aes.new(key: key)
    end

    def run(encrypted_path)
      temp = Tempfile.new([File.basename(encrypted_path), ".tmp"])
      if File.file?(encrypted_path)
        data = read_file(encrypted_path)
        cleartext = @cipher.decrypt(data)
        temp.write(cleartext)
        temp.flush
      end
      temp.close(false)

      unless system("#{@editor} #{Shellwords.escape(temp.path)}")
        raise SymmetricFile::EditError, "User aborted edit"
      end
      new_cleartext = read_file(temp.path)
      if cleartext != new_cleartext
        File.open(encrypted_path, "wb") do |io|
          io.write @cipher.encrypt(new_cleartext)
        end
      end
    ensure
      temp.close! if temp
    end

    private

    def read_file(path)
      File.binread(path)
    rescue Errno::ENOENT
      raise SymmetricFile::InputError, "file '#{path}' not found"
    end
  end
end


lib/symmetric_file/merge_command.rb
require 'shellwords'

module SymmetricFile
  class MergeCommand
    def initialize(key: nil)
      @cipher = SymmetricFile::Aes.new(key: key)
    end

    def run(mine_path, old_path, yours_path)
      mine = Tempfile.open('mine')
      old = Tempfile.open('old')
      yours = Tempfile.open('yours')

      begin
        # decrypt all three of our files
        mine.write @cipher.decrypt(read_file(mine_path))
        old.write @cipher.decrypt(read_file(old_path))
        yours.write @cipher.decrypt(read_file(yours_path))

        # flush our io
        mine.flush
        old.flush
        yours.flush

        cmd = "git merge-file -L mine -L old -L yours -p %s %s %s  2>/dev/null" %
          [ Shellwords.escape(mine.path), Shellwords.escape(old.path), Shellwords.escape(yours.path)]

        diff = `#{cmd}`
        conflict = !$?.success?
      ensure
        # close and unlink our tempfiles
        mine.close!
        old.close!
        yours.close!
      end

      #  create a new tempfile to write our merged file to
      t = Tempfile.open('merged')

      begin
        # encrypt our diff3 output
        t.write @cipher.encrypt(diff)
        t.flush

        # and copy that file back to old_path since that's where git expects to find it
        FileUtils.copy t.path, old_path
      ensure
        t.close!
      end

      # this is important - this exit value is what git uses to decide if there
      # is a conflict or not
      exit (conflict ? 1 : 0)
    end

    private

    def read_file(path)
      File.binread(path)
    rescue Errno::ENOENT
      raise SymmetricFile::InputError, "file '#{path}' not found"
    end
  end
end


lib/symmetric_file/aes.rb
require 'openssl'
require 'digest/sha2'
require 'securerandom'

module SymmetricFile
  # Wrapper for openssl for encrypting and decrypting with AES-256, a symmetric cypher
  # that uses the same key for both encryption and decryption.
  #
  # The encrypted data follows the RNCryptor format, as  described at:
  #
  #   https://github.com/RNCryptor/RNCryptor
  #
  # Core attributes are:
  #
  #  * AES-256 encryption
  #  * CBC mode
  #  * Password stretching with PBKDF2
  #  * Password salting
  #  * Random IV
  #  * Encrypt-then-hash HMAC
  class Aes

    def initialize(key: "1234")
      @key = key
    end

    def encrypt(input)
      version = 3.chr
      options = 1.chr # uses password
      encryption_salt = SecureRandom.random_bytes(8)
      hmac_salt = SecureRandom.random_bytes(8)
      hmac_key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(@key , hmac_salt, 10000, 32)

      cipher = OpenSSL::Cipher.new('aes-256-cbc')
      cipher.encrypt
      cipher.key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(@key, encryption_salt, 10000, 32)
      iv = cipher.random_iv
      ciphertext = cipher.update(input.to_s) + cipher.final
      msg = version + options + encryption_salt + hmac_salt + iv + ciphertext
      hmac = [OpenSSL::HMAC.hexdigest('sha256', hmac_key, msg)].pack('H*')
      msg + hmac
    end

    def decrypt(input)
      raise InputError, "Encrypted data too short" unless input.bytesize >= 67

      version         = input[0,1]
      options         = input[1,1]
      encryption_salt = input[2,8]
      hmac_salt       = input[10,8]
      iv              = input[18,16]
      ciphertext      = input[34, input.bytesize-66]
      hmac            = input[-32, 32]

      raise InputError, "Only version 3 data can be decrypted" unless version == "\x03"

      msg = version + options + encryption_salt + hmac_salt + iv + ciphertext

      hmac_key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(@key, hmac_salt, 10000, 32)
      verified = [OpenSSL::HMAC.hexdigest('sha256', hmac_key, msg)].pack('H*') == hmac

      unless verified
        raise InputError, "HMAC validation failed. The passphrase may be incorrect or the encrypted message may have been tampered with"
      end

      cipher = OpenSSL::Cipher.new('aes-256-cbc')
      cipher.decrypt
      cipher.key = OpenSSL::PKCS5.pbkdf2_hmac_sha1(@key, encryption_salt, 10000, 32)
      cipher.iv = iv
      cipher.update(ciphertext) + cipher.final
    end

  end
end
